피어 2주차 PR

Gwonwoo-Nam opened the pull request:
!

Gwonwoo-Nam committed:
feat:1주차 학습 계획 작성

Gwonwoo-Nam committed:
feat(step1):1단계 모든 Request Header 출력 및 index.html 응답 기능 구현

Gwonwoo-Nam committed:
feat(step-2):QueryParameter 파싱을 위한 HttpRequestUtils 생성 및 HttpRequest 타입 생성

- HttpReqeust 쿼리파라미터와 URL 정보를 저장하는 HttpRequest 타입 구현
- HttpRequestUtils에서 파싱 처리

Gwonwoo-Nam committed:
feat(step2):requestMapping을 위한 UserController 클래스 생성 및 RequestHandler 리팩터링

- RequestHandler에서 mapping - sendResponse로 나누어 수행되도록 메서드 리팩터링
- UserController에서 요청 매핑을 처리해주면 좋을 것 같아 Controller로 매핑 로직을 분리함

honux77 committed:
Merge pull request #3 from Gwonwoo-Nam/study-plan

[poro] 1주차 학습 계획 작성

honux77 committed:
Merge pull request #15 from Gwonwoo-Nam/step1

[poro] step-1 : 모든 Request Header 출력, index.html 뷰 매핑

Gwonwoo-Nam committed:
feat(step2):회원가입 후 홈으로 Redirect 구현

- Controller에서는 Request에 따른 로직을 수행한다.
- RequestHandler에서 리턴된 View에 따라 Mapping하도록 구현

Gwonwoo-Nam committed:
refactor(step-2):HttpRequest에서 쿼리 파라미터를 파싱하여 가지고 있도록 구조 변경

Gwonwoo-Nam committed:
feat(step-2):HttpRequest, HttpRequestUtils, UserController 단위 테스트 작성

Gwonwoo-Nam committed:
feat(step-3):stylesheet 지원 기능 추가 구현

- css, js, fonts 등은 RequestHandler에서 static 파일 경로에서 탐색하도록 수정(html은 templates 디렉토리)
- css는 Content-type을 text/html이 아닌 text/css로 보내도록 변경(Accept 헤더를 참고)

honux77 committed:
Merge pull request #24 from Gwonwoo-Nam/step2

[poro] step2 - 회원가입, 회원가입 후 리다이렉트, 단위 테스트 구현

Gwonwoo-Nam committed:
feat(step3):socket의 연결 상태를 확인하는 기능을 SocketStatusChecker에 구현

- Client 연결 시 Timer 실행, Timer는 정기적으로 ping을 보내 클라이언트 측의 연결 상태를 확인합니다.

Gwonwoo-Nam committed:
feat(step3):Web Client 구현 코드 작성

Gwonwoo-Nam committed:
chore(step-3):클라이언트 측 코드 주석 제거

Gwonwoo-Nam committed:
comment(step-3):주석 작성 및 webclient 테스트 폴더로 경로 이동

Gwonwoo-Nam committed:
feat(step-3):HttpResponse, HttpHeaders 타입 구현하여 응답 리팩토링

- 객체가 메시지를 map 형태로 가지고 있다가 메시지를 보낼 때 toString으로 변환해주는 방식으로 구현
- 자주 사용되고 종류가 많은 ContentsType의 경우 enum 타입으로 생성

Gwonwoo-Nam committed:
refactor(step-3):패키지 경로 재설정

- util에 많은 책임이 있어, request, response 패키지를 생성해 나누었습니다.

honux77 committed:
Merge pull request #36 from Gwonwoo-Nam/step3

[poro] step-3 Stylesheet 지원 및 클라이언트 소켓 구현

Gwonwoo-Nam committed:
refactor(step4):UrlMapping하는 기능과 User 기능을 분리하기 위해 UrlMapper 객체 구현

Gwonwoo-Nam committed:
refactor(step-4):Content-type 객체가 식별자(경로로 구분)와 위치를 가지고있도록 구현하여 리팩토링, HTTP status 코드와 메시지도 status 객체에서 관리

- status enum : 200, 302 리다이렉트
- ContentType : 식별자, content-type, 폴더 경로를 멤버변수로 가진다.

Gwonwoo-Nam committed:
refactor(step-4):View 타입과 View를 매핑해주는 ViewResolver로 기존 RequestHandler에서 처리하던 뷰 매핑과 Response 메시지 전송의 책임을 나눔

- ViewResolver : viewName으로 적절한 view 객체를 생성, 파일경로에서 실제 파일을 매핑해줍니다.
- View : httpResponse 메시지 형태로 렌더링해줍니다.

Gwonwoo-Nam committed:
refactor(step-4):컨텐츠 타입을 경로 식별에서 확장자를 정규표현식으로 식별하도록 변경

Gwonwoo-Nam committed:
chore(step-4):미사용 코드 삭제, 주석 추가

honux77 committed:
Merge pull request #50 from Gwonwoo-Nam/step3-refactor

[poro] step3 피드백 반영 및 리팩터링 : ContentType 개선, UrlMapper, View, ViewResolver

Gwonwoo-Nam committed:
refactor(step-4):ContentsType의 기본 경로 상수로 Refactoring

Gwonwoo-Nam committed:
feat(step-4):회원가입 요청 GET에서 POST 요청으로 변경, URL user/create에서 users/create로 변경

Gwonwoo-Nam committed:
feat(step-4):HttpHeaders가 String을 인자로 받아 맵에 저장하는 parse 메서드 구현, Content Length를 반환하는 getContentLength 구현

- parse(String headerString):헤더를 한줄 씩 읽어 key value 형태로 header 맵에 저장합니다.
- getContentLength() : header 맵에 content-length 키가 존재하면 길이를 반환하고 아니면 0을 리턴합니다.

Gwonwoo-Nam committed:
feat(step-4):리다이렉트 response 반환 기능 수정

- Redirect는 body가 없는데 분기문으로 나누어, Body를 제외하고 메시지를 render하도록 수정

Gwonwoo-Nam committed:
feat(step-4):회원가입 Post 요청으로 처리하도록 RequestHandler run 메소드 수정

- Content가 있는 경우, 없는 경우로 나누어서 body가 있으면 body 필드에 저장

Gwonwoo-Nam committed:
refactor(step-4):response가 view 객체를 참조하도록 구조 변경

- response가 메시지를 스스로 작성하도록 로직 변경, View는 더 이상 response를 참조하지 않는다.

honux77 committed:
Merge pull request #66 from Gwonwoo-Nam/step4

[poro] Step-4 POST로 회원가입

Gwonwoo-Nam committed:
feat(step-5):Controller 구조 Front Controller와 상세 Controller로 구조 분리

- UrlMapper 삭제, Controller 인터페이스 구현 및 FrontController 생성(기존 DispatcherServlet 역할과 유사)

Gwonwoo-Nam committed:
feat(step-5):Controller 구조 Front Controller와 상세 Controller로 구조 분리

- UrlMapper 삭제, Controller 인터페이스 구현 및 FrontController 생성(기존 DispatcherServlet 역할과 유사)

Gwonwoo-Nam committed:
test(step-5):테스트코드 메서드 이름 변경

Gwonwoo-Nam committed:
feat(step-5):Reflection을 이용해 @RequestMapping 파싱하여 URL을 매핑시키는 기능 구현

- URL마다 하나의 컨트롤러가 할당됩니다. (서블릿과 비슷한 구조)
- Reflection으로 Controller의 RequestMapping Annotation의 URL 정보를 읽어와 Map으로 저장합니다.(URL, Controller)의 key value 구조.

Gwonwoo-Nam committed:
refactor(step-5):홈 URL 컨트롤러 ArticleController -> HomeController로 변경

Gwonwoo-Nam committed:
feat(step-5):HttpResponse를 Controller에서도 인자를 가져서 접근이 가능하도록 의존성 수정

- Response를 미리 만들어서 Controller에 주입시켜주는 구조로 변경함

Gwonwoo-Nam committed:
refactor(step-5):HttpResponse Builder 패턴으로 생성하도록 리팩터링(가독성 개선)

Gwonwoo-Nam committed:
feat(step-5):UserLoginController 구현, users/login url 매핑

- 세션, 쿠키 기능 구현 예정
- user/login -> users/login으로 변경

jaea-kim commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,50 +1,89 @@
-            // TODO 사용자 요청에 대한 처리는 이 곳에 구현하면 된다.
+            BufferedReader br = new BufferedReader(new InputStreamReader(in, "UTF-8"));
+
+            HttpRequest httpRequest = setupHttpRequest(br);
```
http setup을 request에서 하지 않고 requestHandler 에서 한 이유가 있나요?

Gwonwoo-Nam replied:
예리한 지적이십니다..

honux77 commented on build.gradle:
```diff
@@ -22,6 +22,7 @@ dependencies {
     testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
     testImplementation 'org.assertj:assertj-core:3.16.1'
 
```
mockito 의존성이 빠져 있군요!

jaea-kim reviewed (commented):
미천한.. 제가 .. 감히 어찌 스프링을 평가하겠습니까........ 갓포로 귀한 곳에 잠시나마 들어올 수 있어 평생의 축복입니다 갓포로 코드읽기 하차합니다...

Gwonwoo-Nam committed:
test(step-5):mock 의존성 추가, 테스트 코드 수정

- Reflection 시 경로에 의존해서 테스트 코드에서는 제대로 동작이 안된다. 어떻게 해결해야할지 고민중입니다.

Gwonwoo-Nam committed:
refactor(step-5):FrontController의 다룰 handler를 mapping 기능 HandlerMapper 클래스로 분리

- HandlerMapper는 static으로 map 기능 제공

Gwonwoo-Nam committed:
feat(step-5):Custom Annotation으로 컨트롤러에 Method 타입을 명시하는 기능 구현

- Method 타입 표시 시 컨트롤러 생성 시에 reflection으로 annotation이 붙은 메소드를 읽어들여 실행 map에 등록한다.

Gwonwoo-Nam committed:
refactor(step-5):HandlerMapping 초기화 시에만 발생하도록 리팩토링

- 기존은 요청을 처리할 때마다 mapping이 발생해서 비효율적이었다.

Gwonwoo-Nam committed:
chore:test 컴파일 오류 재수정

honux77 committed:
Merge pull request #70 from Gwonwoo-Nam/step5

[poro] step 4 컨트롤러 리팩터링, Annotation 자동 URL 매핑 구현

Gwonwoo-Nam committed:
refactor(step-5):controller interface에서 추상 클래스로 변경

- 컨트롤러가 기본 메서드를 상속받도록 변경(default 메소드로 상속과 동일하게 사용한다고 판단해서 변경했습니다)

Gwonwoo-Nam committed:
feat(step-5):로그인 시 쿠키와 쿠키 정보를 저장하는 메모리 SessionDB 구현

Gwonwoo-Nam committed:
refactor(step-5):httpRequest 생성 HttpRequestBuilder 클래스를 생성하여 위임

Gwonwoo-Nam committed:
refactor(step-5):session 패키지로 세션 관련 기능 패키지 변경

Gwonwoo-Nam committed:
feat(step-5):ExceptionHandler 기능 구현

- 던져지는 예외 타입에 따라 처리하도록 구현하고 싶었으나, Invocation Exception으로 던져져서 식별 불가
- Exception 발생 시 ExceptionHandler가 실행되고, Handler가 없을 때는 예외를 다시 던지도록 구현 완료

Gwonwoo-Nam committed:
feat(step-5):ExceptionHandler 기능 개선 완료

- 학습 후 Invocation Exception이 exception을 wrapping하고 있다는 것을 알게되어, invoke한 메서드에서 exception을 가져와서 어노테이션의 exception과 비교 후 예외 처리 로직을 호출하도록 개선했습니다.
- 초기화에 사용되는 reflection mapper 클래스들을 mapper 패키지로 분리했습니다.

honux77 commented on src/main/java/session/SessionDb.java:
```diff
@@ -0,0 +1,18 @@
+    public static String addSessionedUser(String userId) {
+        Session userSession = new Session();
+        sessionMap.put(userId, userSession);
+        return userSession.getSessionId();
```
이렇게도 동작할 수 있긴 하겠네요. 보통은 K: SessionId로 구현하는게 일반적이에요.
